module/zfs/zfs_log.c

```diff
@@ -820,5 +821,56 @@ zfs_log_acl(zilog_t *zilog, dmu_tx_t *tx, znode_t *zp,
 	zil_itx_assign(zilog, itx, tx);
 }
 
+/*
+ * Handles TX_CLONE transactions.
+ */
+void
+zfs_log_clone(zilog_t *zilog, dmu_tx_t *tx, int txtype, znode_t *zp,
+    int ioflag, uint64_t offset, uint64_t length, uint64_t blksz,
+    const blkptr_t *bps, size_t nbps)
+{
+	itx_t *itx;
+	lr_clone_t *lr;
+	uint64_t len, max_log_data;
+	size_t i, partnbps;
+
+	if (zil_replaying(zilog, tx) || zp->z_unlinked)
+	    /* XXX: zfs_xattr_owner_unlinked(zp)? */
+		return;
+
+	max_log_data = zil_max_log_data(zilog, sizeof (lr_clone_t));
+
+	do {
+		partnbps = MIN(nbps, max_log_data / sizeof (bps[0]));
+		len = 0;
+		for (i = 0; i < partnbps; i++) {
+			len += BP_GET_LSIZE(&bps[i]);
+		}
+		len = MIN(len, length);
+
+		itx = zil_itx_create(txtype,
+		    sizeof (*lr) + sizeof (bps[0]) * partnbps);
+		lr = (lr_clone_t *)&itx->itx_lr;
+		lr->lr_foid = zp->z_id;
+		lr->lr_offset = offset;
+		lr->lr_length = len;
+		lr->lr_blksz = blksz;
+		lr->lr_nbps = partnbps;
+		memcpy(lr->lr_bps, bps, sizeof (bps[0]) * partnbps);
```

Comment:
Do we have some guaranties that blocks from different dataset won't be freed before the replay?  For example in case of encrypted dataset (I just don't remember whether non-encrypted datasets are replayed strictly on import or may be sometimes later).

Reply:
IIRC, they are replayed at mount, so it could be late

Reply:
For now I changed the code to only use ZIL when we clone within a single dataset. When sync=always is set on the destination dataset and we are cloning between separate datasets we call txg_wait_synced() before returning.

Reply:
I suspect unless the file is really big it may be faster to return error and just copy it than wait for TXG commit.